File: gurobi/multicommodity/multicommodity.cpp
#include "gurobi_c++.h"

#include <iostream>
#include <fstream>
#include <cstring>
#include <cmath>
#include <vector>

using namespace std;

int main(int argc, char *argv[]) {
    try {
        // input data
        ifstream data("multicommodity.dat");
        
        size_t norig = 0;
        size_t ndest = 0;
        size_t nprod = 0;
        
        data >> norig;
        data >> ndest;
        data >> nprod;
        
        double **supply = new double *[norig];
        for (size_t i = 0; i < norig; ++i)
            supply[i] = new double [nprod];
        
        double **demand = new double *[ndest];
        for (size_t j = 0; j < ndest; ++j)
            demand[j] = new double [nprod];
        
        double **limit = new double *[norig];
        for (size_t i = 0; i < norig; ++i)
            limit[i] = new double [ndest];
        
        double ***cost = new double **[norig];
        for (size_t i = 0; i < norig; ++i) {
            cost[i] = new double *[ndest];
            
            for (size_t j = 0; j < ndest; ++j)
                cost[i][j] = new double [nprod];
        }
        
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < nprod; ++j)
                data >> supply[i][j];
        
        for (size_t j = 0; j < ndest; ++j)
            for (size_t k = 0; k < nprod; ++k)
                data >> demand[j][k];
        
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < ndest; ++j)
                data >> limit[i][j];
        
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < ndest; ++j)
                for (size_t k = 0; k < nprod; ++k)
                    data >> cost[i][j][k];
        
        data.close();
        
        // environment and models
        GRBEnv env = GRBEnv();
        GRBModel master = GRBModel(env);
        GRBModel sub = GRBModel(env);
        
        // disable log information
        master.set(GRB_IntParam_OutputFlag, 0);
        sub.set(GRB_IntParam_OutputFlag, 0);
        
        // iteration count
        size_t iter = 0;
        
        // variables, constraints and parameters in mater problem
        GRBConstr **multi = new GRBConstr *[norig];
        for (size_t i = 0; i < norig; ++i)
            multi[i] = new GRBConstr [ndest];
        
        GRBConstr *convex = NULL;
        
        vector<GRBVar> weight;
        GRBVar excess;
        
        GRBLinExpr obj_masteri = 0.0;
        GRBLinExpr obj_masterii = 0.0;
        GRBLinExpr obj_masteriii = 0.0;
        
        GRBColumn col;
        
        double **xtrans = NULL;
        double sum_xtrans = 0.0;
        double sum_costxtrans = 0.0;
        vector<double> propcost;
        
        // variables and parameters in sub problem
        GRBVar ***trans = new GRBVar **[norig];
        for (size_t i = 0; i < norig; ++i) {
            trans[i] = new GRBVar *[ndest];
            
            for (size_t j = 0; j < ndest; ++j)
                trans[i][j] = new GRBVar [nprod];
        }
        
        GRBLinExpr obj_subi = 0.0;
        GRBLinExpr obj_subii = 0.0;
        
        GRBLinExpr con_supply = 0.0;
        GRBLinExpr con_demand = 0.0;
        
        GRBLinExpr con_optmulti = 0.0;
        
        double priceconvex = 1.0;
        double **price = new double *[norig];
        for (size_t i = 0; i < norig; ++i)
            price[i] = new double [ndest];
        
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < ndest; ++j)
                price[i][j] = 0.0;
        
        double **optship = new double *[norig];
        for (size_t i = 0; i < norig; ++i)
            optship[i] = new double [ndest];
        
        // add variable 'excess' for master problem
        excess = master.addVar(0.0, GRB_INFINITY, 0.0, GRB_CONTINUOUS, string("excess"));
        
        // add temporary constraints 'multi' for master problem
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < ndest; ++j)
                multi[i][j] = master.addConstr(-excess <= limit[i][j], "multi" + to_string(i) + to_string(j));
        
        // add temporary constraint 'convex' for master problem
        convex = master.addConstrs(1);
        convex[0].set(GRB_CharAttr_Sense, GRB_EQUAL);
        convex[0].set(GRB_DoubleAttr_RHS, 1.0);
        convex[0].set(GRB_StringAttr_ConstrName, string("convex"));
        
        // set objective for master problem in 'Phase I'
        obj_masteri = excess;
        
        // add variables 'trans' for subproblem
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < ndest; ++j)
                for (size_t k = 0; k < nprod; ++k)
                    trans[i][j][k] = sub.addVar(0.0, GRB_INFINITY, 0.0, GRB_CONTINUOUS, "trans" + to_string(i) + to_string(j) + to_string(k));
        
        // add constraints 'supply' for subproblem
        for (size_t i = 0; i < norig; ++i) {
            for (size_t k = 0; k < nprod; ++k) {
                for (size_t j = 0; j < ndest; ++j)
                    con_supply += trans[i][j][k];
                
                sub.addConstr(con_supply == supply[i][k], "supply" + to_string(i) + to_string(k));
                con_supply = 0.0;
            }
        }
        
        // add constraints 'demand' for subproblem
        for (size_t j = 0; j < ndest; ++j) {
            for (size_t k = 0; k < nprod; ++k) {
                for (size_t i = 0; i < norig; ++i)
                    con_demand += trans[i][j][k];
                
                sub.addConstr(con_demand == demand[j][k], "demand" + to_string(j) + to_string(k));
                con_demand = 0.0;
            }
        }
        
        // set objective for subproblem in 'Phase I'
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < ndest; ++j)
                for (size_t k = 0; k < nprod; ++k)
                    obj_subi -= price[i][j] * trans[i][j][k];
        
        obj_subi -= priceconvex;
        
        // dantzig-wolfe decomposition
        cout << "               *** Dantzig-Wolfe Decomposition ***               " << endl;
        cout << endl;
        
        // set objective for subproblem in 'Phase I'
        sub.setObjective(obj_subi, GRB_MINIMIZE);
        
        // set objective for master problem in 'Phase I'
        master.setObjective(obj_masteri, GRB_MINIMIZE);
        
        // 'Phase I' of dantzig-wolfe decomposition
        cout << "Phase I: " << endl;
        for (;;) {
            cout << "Iteration: " << iter << endl;
            
            // solve subproblem in 'Phase I'
            sub.optimize();
            
            if (sub.get(GRB_DoubleAttr_ObjVal) >= -1e-6) {
                cout << "No feasible solution..." << endl;
                break;
            }
            else {
                ++iter;
                
                // calculate parameters for master problem
                xtrans = new double *[norig];
                for (size_t i = 0; i < norig; ++i)
                    xtrans[i] = new double [ndest];
                
                sum_xtrans = 0.0;
                for (size_t i = 0; i < norig; ++i) {
                    for (size_t j = 0; j < ndest; ++j) {
                        for (size_t k = 0; k < nprod; ++k)
                            sum_xtrans += trans[i][j][k].get(GRB_DoubleAttr_X);
                        
                        xtrans[i][j] = sum_xtrans;
                        sum_xtrans = 0.0;
                    }
                }
                
                sum_costxtrans = 0.0;
                for (size_t i = 0; i < norig; ++i)
                    for (size_t j = 0; j < ndest; ++j)
                        for (size_t k = 0; k < nprod; ++k)
                            sum_costxtrans += cost[i][j][k] * trans[i][j][k].get(GRB_DoubleAttr_X);
                
                propcost.push_back(sum_costxtrans);
                
                // update constraints in master problem
                col = GRBColumn();
                for (size_t i = 0; i < norig; ++i)
                    for (size_t j = 0; j < ndest; ++j)
                        col.addTerm(xtrans[i][j], multi[i][j]);
                
                col.addTerm(1.0, convex[0]);
                
                // add variable 'weight'
                weight.push_back(master.addVar(0.0, GRB_INFINITY, 0.0, GRB_CONTINUOUS, col, "weight" + to_string(iter)));
                
                // deallocate heap memory
                for (size_t i = 0; i < norig; ++i)
                    delete [] xtrans[i];
                delete [] xtrans;
                
                // solve master problem in 'Phase I'
                master.optimize();
                
                // update price
                if (master.get(GRB_DoubleAttr_ObjVal) <= 1e-6)
                    break;
                else {
                    for (size_t i = 0; i < norig; ++i)
                        for (size_t j = 0; j < ndest; ++j)
                            price[i][j] = multi[i][j].get(GRB_DoubleAttr_Pi);
                    
                    priceconvex = convex[0].get(GRB_DoubleAttr_Pi);
                }
                
                // reset objective for subproblem in 'Phase I'
                obj_subi = 0.0;
                for (size_t i = 0; i < norig; ++i)
                    for (size_t j = 0; j < ndest; ++j)
                        for (size_t k = 0; k < nprod; ++k)
                            obj_subi -= price[i][j] * trans[i][j][k];
                
                obj_subi -= priceconvex;
                
                sub.setObjective(obj_subi, GRB_MINIMIZE);
            }
        }
        
        // setting up for 'Phase II'
        cout << "Setting up for Phase II..." << endl;
        
        // set objective for master problem in 'Phase II'
        for (size_t i = 0; i < weight.size(); ++i)
            obj_masterii += propcost[i] * weight[i];
        
        master.setObjective(obj_masterii, GRB_MINIMIZE);
        
        // fix variable 'excess'
        double xexcess = excess.get(GRB_DoubleAttr_X);
        
        excess.set(GRB_DoubleAttr_LB, xexcess);
        excess.set(GRB_DoubleAttr_UB, xexcess);
        
        // solve master problem in 'Phase II'
        master.optimize();
        
        // update price
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < ndest; ++j)
                price[i][j] = multi[i][j].get(GRB_DoubleAttr_Pi);
        
        priceconvex = convex[0].get(GRB_DoubleAttr_Pi);
        
        // set objective for subproblem in 'Phase II'
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < ndest; ++j)
                for (size_t k = 0; k < nprod; ++k)
                    obj_subii += (cost[i][j][k] - price[i][j]) * trans[i][j][k];
        
        obj_subii -= priceconvex;
        
        sub.setObjective(obj_subii, GRB_MINIMIZE);
        
        // increase iteration count
        ++iter;
        
        // 'Phase II' of dantzig-wolfe decomposition
        cout << "Phase II: " << endl;
        for (;;) {
            cout << "Iteration: " << iter << endl;
            
            // solve subproblem in 'Phase II'
            sub.optimize();
            
            if (sub.get(GRB_DoubleAttr_ObjVal) >= -1e-6) {
                cout << "Optimal solution..." << endl;
                break;
            }
            else {
                ++iter;
                
                // calculate parameters for master problem
                xtrans = new double *[norig];
                for (size_t i = 0; i < norig; ++i)
                    xtrans[i] = new double [ndest];
                
                sum_xtrans = 0.0;
                for (size_t i = 0; i < norig; ++i) {
                    for (size_t j = 0; j < ndest; ++j) {
                        for (size_t k = 0; k < nprod; ++k)
                            sum_xtrans += trans[i][j][k].get(GRB_DoubleAttr_X);
                        
                        xtrans[i][j] = sum_xtrans;
                        sum_xtrans = 0.0;
                    }
                }
                
                sum_costxtrans = 0.0;
                for (size_t i = 0; i < norig; ++i)
                    for (size_t j = 0; j < ndest; ++j)
                        for (size_t k = 0; k < nprod; ++k)
                            sum_costxtrans += cost[i][j][k] * trans[i][j][k].get(GRB_DoubleAttr_X);
                
                // update constraints in master problem
                col = GRBColumn();
                for (size_t i = 0; i < norig; ++i)
                    for (size_t j = 0; j < ndest; ++j)
                        col.addTerm(xtrans[i][j], multi[i][j]);
                
                col.addTerm(1.0, convex[0]);
                
                // add variable 'weight'
                weight.push_back(master.addVar(0.0, GRB_INFINITY, sum_costxtrans, GRB_CONTINUOUS, col, "weight" + to_string(iter)));
                
                // deallocate heap memory
                for (size_t i = 0; i < norig; ++i)
                    delete [] xtrans[i];
                delete [] xtrans;
                
                // solve master problem in 'Phase II'
                master.optimize();
                
                // update price
                for (size_t i = 0; i < norig; ++i)
                    for (size_t j = 0; j < ndest; ++j)
                        price[i][j] = multi[i][j].get(GRB_DoubleAttr_Pi);
                    
                priceconvex = convex[0].get(GRB_DoubleAttr_Pi);
                
                // set objective for subproblem in 'Phase II'
                obj_subii = 0.0;
                for (size_t i = 0; i < norig; ++i)
                    for (size_t j = 0; j < ndest; ++j)
                        for (size_t k = 0; k < nprod; ++k)
                            obj_subii += (cost[i][j][k] - price[i][j]) * trans[i][j][k];
                
                obj_subii -= priceconvex;
                
                sub.setObjective(obj_subii, GRB_MINIMIZE);
            }
        }

        // 'Phase III' of dantzig-wolfe decomposition
        cout << "Phase III: " << endl;
        
        // set objective for master problem in 'Phase III'
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < ndest; ++j)
                optship[i][j] = limit[i][j] + excess.get(GRB_DoubleAttr_X) - multi[i][j].get(GRB_DoubleAttr_Slack);
        
        for (size_t i = 0; i < norig; ++i)
            for (size_t j = 0; j < ndest; ++j)
                for (size_t k = 0; k < nprod; ++k)
                    obj_masteriii += cost[i][j][k] * trans[i][j][k];
        
        sub.setObjective(obj_masteriii, GRB_MINIMIZE);
        
        for (size_t i = 0; i < norig; ++i) {
            for (size_t j = 0; j < ndest; ++j) {
                for (size_t k = 0; k < nprod; ++k)
                    con_optmulti += trans[i][j][k];
                
                sub.addConstr(con_optmulti == optship[i][j], "optmulti" + to_string(i) + to_string(j));
                con_optmulti = 0.0;
            }
        }
        
        // solve master problem in 'Phase III'
        sub.optimize();
        
        // end loop
        cout << "               *** End Loop ***               " << endl;
        
        // report solution
        cout << "               *** Summary Report ***               " << endl;
        cout << "Objective: " << sub.get(GRB_DoubleAttr_ObjVal) << endl;
        cout << "Variables: " << endl;
        for (size_t i = 0; i < norig; ++i) {
            for (size_t j = 0; j < ndest; ++j) {
                for (size_t k = 0; k < nprod; ++k) {
                    if (fabs(trans[i][j][k].get(GRB_DoubleAttr_X)) >= 1e-6)
                        printf("  trans[%d][%d][%d] = %12.6f\n", i, j, k, trans[i][j][k].get(GRB_DoubleAttr_X));
                }
            }
        }
        
        // deallocate heap memory
        for (size_t i = 0; i < norig; ++i)
            delete [] supply[i];
        delete [] supply;
        
        for (size_t i = 0; i < ndest; ++i)
            delete [] demand[i];
        delete [] demand;
        
        for (size_t i = 0; i < norig; ++i)
            delete [] limit[i];
        delete [] limit;
        
        for (size_t i = 0; i < norig; ++i) {
            for (size_t j = 0; j < ndest; ++j)
                delete [] cost[i][j];
            delete [] cost[i];
        }
        delete [] cost;
        
        for (size_t i = 0; i < norig; ++i)
            delete [] multi[i];
        delete [] multi;
        
        for (size_t i = 0; i < norig; ++i) {
            for (size_t j = 0; j < ndest; ++j)
                delete [] trans[i][j];
            delete [] trans[i];
        }
        delete [] trans;
        
        for (size_t i = 0; i < norig; ++i)
            delete [] price[i];
        delete [] price;
        
        for (size_t i = 0; i < norig; ++i)
            delete [] optship[i];
        delete [] optship;
    }
    catch (GRBException e) {
        cout << "Error code = " << e.getErrorCode() << endl;
        cout << e.getMessage() << endl;
    }
    catch(...) {
        cout << "Exception during optimization" << endl;
    }
    
    return 0;
}

File: gurobi/multicommodity/multicommodity.py
from __future__ import division, print_function

import gurobipy as GRBPY


class Multicommodity:
    def __init__(self):
        # initialize data
        self.norig = 0
        self.ndest = 0
        self.nprod = 0
        self.supply = []
        self.demand = []
        self.limit = []
        self.cost = []

        # initialize variables and constraints
        self.cmulti = []
        self.vtrans = []
        self.vweight = []

        # initialize parameters
        self.iter = 0
        self.priceconvex = 1.0
        self.price = []
        self.propcost = []
    
    def read(self, filename):
        # input data
        with open(filename, "r") as data:
            self.norig = int(data.readline())
            self.ndest = int(data.readline())
            self.nprod = int(data.readline())

            for i in range(self.norig):
                column = data.readline().split()
                lsupply = []
                for k in range(self.nprod):
                    lsupply.append(float(column[k]))
                self.supply.append(lsupply)

            for j in range(self.ndest):
                column = data.readline().split()
                ldemand = []
                for k in range(self.nprod):
                    ldemand.append(float(column[k]))
                self.demand.append(ldemand)

            for i in range(self.norig):
                column = data.readline().split()
                llimit = []
                for j in range(self.ndest):
                    llimit.append(float(column[j]))
                self.limit.append(llimit)

            for i in range(self.norig):
                lcost = []
                for j in range(self.ndest):
                    column = data.readline().split()
                    llcost = []
                    for k in range(self.nprod):
                        llcost.append(float(column[k]))
                    lcost.append(llcost)
                self.cost.append(lcost)
    
    def build(self):
        try:
            # define models for 'master' and 'sub'
            self.master = GRBPY.Model("master")
            self.sub = GRBPY.Model("sub")

            # disable log information
            self.master.setParam("OutputFlag", 0)
            self.sub.setParam("OutputFlag", 0)

            # add variable 'excess' for master problem
            self.vexcess = self.master.addVar(0.0, GRBPY.GRB.INFINITY, 0.0, GRBPY.GRB.CONTINUOUS)

            # add temporary constraints 'multi' for master problem
            for i in range(self.norig):
                lmulti = []
                for j in range(self.ndest):
                    lmulti.append(self.master.addConstr(-self.vexcess <= self.limit[i][j]))
                self.cmulti.append(lmulti)
            
            # tricky approach to add temporary constraint 'convex' for master problem
            self.cconvex = self.master.addConstr(1e-6 * self.vexcess == 1.0)
            
            # add variables 'trans' for subproblem
            for i in range(self.norig):
                ltrans = []
                for j in range(self.ndest):
                    lltrans = []
                    for k in range(self.nprod):
                        lltrans.append(self.sub.addVar(0.0, GRBPY.GRB.INFINITY, 0.0, GRBPY.GRB.CONTINUOUS))
                    ltrans.append(lltrans)
                self.vtrans.append(ltrans)
            
            # add constraints 'supply' for subproblem
            for i in range(self.norig):
                for k in range(self.nprod):
                    self.sub.addConstr(GRBPY.quicksum(self.vtrans[i][j][k] for j in range(self.ndest)) \
                                                      == self.supply[i][k])
            
            # add constraints 'demand' for subproblem
            for j in range(self.ndest):
                for k in range(self.nprod):
                    self.sub.addConstr(GRBPY.quicksum(self.vtrans[i][j][k] for i in range(self.norig)) \
                                                      == self.demand[j][k])
        except GRBPY.GurobiError as e:
            print('Error code' + str(e.errno) + ': ' + str(e))
        except AttributeError as e:
            print('Encountered an attribute error: ' + str(e))
    
    def solvePhaseI(self):
        # initialize parameters
        for i in range(self.norig):
            lprice = [0.0] * self.ndest
            self.price.append(lprice)
        
        obj_masteri = self.vexcess
        obj_subi = -GRBPY.quicksum(self.price[i][j] * self.vtrans[i][j][k] \
                                   for i in range(self.norig) \
                                   for j in range(self.ndest) \
                                   for k in range(self.nprod)) - self.priceconvex

        # set objective for master problem in 'Phase I'
        self.master.setObjective(obj_masteri, GRBPY.GRB.MINIMIZE)

        # set objective for subproblem in 'Phase I'
        self.sub.setObjective(obj_subi, GRBPY.GRB.MINIMIZE)

        # ok, forget this
        self.master.chgCoeff(self.cconvex, self.vexcess, 0.0)

        # 'Phase I' of dantzig-wolfe decomposition
        print("Phase I: ")

        while True:
            print("Iteration: ", self.iter)

            # solve subproblem in 'Phase I'
            self.sub.optimize()

            if self.sub.objval >= -1e-6:
                print("No feasible solution...")
                break
            else:
                self.iter += 1

                # calculate parameters for master problem
                sum_costxtrans = sum(self.cost[i][j][k] * self.vtrans[i][j][k].x \
                                     for i in range(self.norig) \
                                     for j in range(self.ndest) \
                                     for k in range(self.nprod))

                self.propcost.append(sum_costxtrans)

                # update constraints in master problem
                col = GRBPY.Column()
                for i in range(self.norig):
                    for j in range(self.ndest):
                        col.addTerms(sum(self.vtrans[i][j][k].x for k in range(self.nprod)), self.cmulti[i][j])
                
                col.addTerms(1.0, self.cconvex)

                # add variable 'weight'
                self.vweight.append(self.master.addVar(0.0, GRBPY.GRB.INFINITY, 0.0, GRBPY.GRB.CONTINUOUS, "", col))

                # solve master problem in 'Phase I'
                self.master.optimize()

                # update price
                if self.master.objval <= 1e-6:
                    break
                else:
                    for i in range(self.norig):
                        for j in range(self.ndest):
                            self.price[i][j] = self.cmulti[i][j].pi
                    
                    self.priceconvex = self.cconvex.pi
                
                # reset objective for subproblem in 'Phase I'
                obj_subi = -GRBPY.quicksum(self.price[i][j] * self.vtrans[i][j][k] \
                                           for i in range(self.norig) \
                                           for j in range(self.ndest) \
                                           for k in range(self.nprod)) - self.priceconvex

                self.sub.setObjective(obj_subi, GRBPY.GRB.MINIMIZE)
    
    def setupPhaseII(self):
        # setting up for 'Phase II'
        print("Setting up for Phase II...")

        # set objective for master problem in 'Phase II'
        obj_masterii = GRBPY.quicksum(self.propcost[i] * self.vweight[i] for i in range(len(self.propcost)))

        self.master.setObjective(obj_masterii, GRBPY.GRB.MINIMIZE)

        # fix variable 'excess'
        self.vexcess.lb = self.vexcess.x
        self.vexcess.ub = self.vexcess.x

        # solve master problem in 'Phase II'
        self.master.optimize()

        # update price
        for i in range(self.norig):
            for j in range(self.ndest):
                self.price[i][j] = self.cmulti[i][j].pi
        
        self.priceconvex = self.cconvex.pi

        # set objective for subproblem in 'Phase II'
        obj_subii = GRBPY.quicksum((self.cost[i][j][k] - self.price[i][j]) * self.vtrans[i][j][k] \
                                   for i in range(self.norig) \
                                   for j in range(self.ndest) \
                                   for k in range(self.nprod)) - self.priceconvex
        
        self.sub.setObjective(obj_subii, GRBPY.GRB.MINIMIZE)

        # increase iteration count
        self.iter += 1

    def solvePhaseII(self):
        # 'Phase II' of dantzig-wolfe decomposition
        print("Phase II: ")

        while True:
            print("Iteration: ", self.iter)

            # solve subproblem in 'Phase II'
            self.sub.optimize()

            if self.sub.objval >= -1e-6:
                print("Optimal solution...")
                break
            else:
                self.iter += 1

                # calculate parameters for master problem
                sum_costxtrans = sum(self.cost[i][j][k] * self.vtrans[i][j][k].x \
                                     for i in range(self.norig) \
                                     for j in range(self.ndest) \
                                     for k in range(self.nprod))
                
                # update constraints in master problem
                col = GRBPY.Column()
                for i in range(self.norig):
                    for j in range(self.ndest):
                        col.addTerms(sum(self.vtrans[i][j][k].x for k in range(self.nprod)), self.cmulti[i][j])
                
                col.addTerms(1.0, self.cconvex)

                # add variable 'weight'
                self.vweight.append(self.master.addVar(0.0, GRBPY.GRB.INFINITY, sum_costxtrans, GRBPY.GRB.CONTINUOUS, "", col))

                # solve master problem in 'Phase II'
                self.master.optimize()

                # update price
                for i in range(self.norig):
                    for j in range(self.ndest):
                        self.price[i][j] = self.cmulti[i][j].pi
                
                self.priceconvex = self.cconvex.pi

                # reset objective for subproblem in 'Phase II'
                obj_subii = GRBPY.quicksum((self.cost[i][j][k] - self.price[i][j]) * self.vtrans[i][j][k] \
                                            for i in range(self.norig) \
                                            for j in range(self.ndest) \
                                            for k in range(self.nprod)) - self.priceconvex
                
                self.sub.setObjective(obj_subii, GRBPY.GRB.MINIMIZE)
    
    def solvePhaseIII(self):
        optship = []
        for i in range(self.norig):
            loptship = [0.0] * self.ndest
            optship.append(loptship)
        
        # 'Phase III' of dantzig-wolfe decomposition
        print("Phase III:")

        # set objective for master problem in 'Phase III'
        for i in range(self.norig):
            for j in range(self.ndest):
                optship[i][j] = self.limit[i][j] + self.vexcess.x - self.cmulti[i][j].slack
        
        obj_masteriii = GRBPY.quicksum(self.cost[i][j][k] * self.vtrans[i][j][k] \
                                       for i in range(self.norig) \
                                       for j in range(self.ndest) \
                                       for k in range(self.nprod))
        
        self.sub.setObjective(obj_masteriii, GRBPY.GRB.MINIMIZE)

        for i in range(self.norig):
            for j in range(self.ndest):
                self.sub.addConstr(GRBPY.quicksum(self.vtrans[i][j][k] for k in range(self.nprod)) == optship[i][j])
        
        # solve master problem in 'Phase III'
        self.sub.optimize()

    def solve(self):
        # build 'master' and 'sub'
        self.build()

        # dantzig-wolfe decomposition
        print("               *** Dantzig-Wolfe Decomposition ***               ")
        self.solvePhaseI()

        self.setupPhaseII()

        self.solvePhaseII()

        self.solvePhaseIII()
        print("                        *** End Loop ***                        ")
    
    def report(self):
        # report solution
        print("               *** Summary Report ***                           ")
        print("Objective: ", self.sub.objval)
        print("Variables: ")
        for i in range(self.norig):
            for j in range(self.ndest):
                for k in range(self.nprod):
                    if abs(self.vtrans[i][j][k].x) >= 1e-6:
                        print("  trans[%d][%d][%d] = %12.6f" % (i, j, k, self.vtrans[i][j][k].x))

if __name__ == "__main__":
    multicomm = Multicommodity()
    multicomm.read("multicommodity.dat")
    multicomm.solve()
    multicomm.report()